src/core/lib/iomgr/ev_epollsig_linux.c

```diff
@@ -1657,6 +1695,128 @@ bool grpc_are_polling_islands_equal(void *p, void *q) {
 }
 
 /*******************************************************************************
+ * Engine Postfork Handlers
+ */
+
+/* Only used with FORK_SUPPORT_ENABLED=1 */
+static gpr_mu g_fd_mu;
+static grpc_fd *g_root_fd = NULL;
+
+static gpr_mu g_island_mu;
+static struct polling_island *g_root_island = NULL;
+
+static void global_fork_fd_list_add(grpc_fd *fd) {
+  if (grpc_fork_support_enabled()) {
+    gpr_mu_lock(&g_fd_mu);
+    fd->global_prev = NULL;
+    fd->global_next = g_root_fd;
+    if (g_root_fd) {
+      g_root_fd->global_prev = fd;
+    }
+    g_root_fd = fd;
+    gpr_mu_unlock(&g_fd_mu);
+  }
+}
+
+static void global_fork_fd_list_remove(grpc_fd *fd) {
+  if (grpc_fork_support_enabled()) {
+    gpr_mu_lock(&g_fd_mu);
+    if (fd->global_prev) {
+      fd->global_prev->global_next = fd->global_next;
+    }
+    if (fd->global_next) {
+      fd->global_next->global_prev = fd->global_prev;
+    }
+    if (g_root_fd == fd) {
+      g_root_fd = g_root_fd->global_next;
+    }
+    gpr_mu_unlock(&g_fd_mu);
+  }
+}
+
+static void global_fork_pi_list_add(polling_island *pi) {
+  if (grpc_fork_support_enabled()) {
+    gpr_mu_lock(&g_island_mu);
+    pi->prev = NULL;
+    pi->next = g_root_island;
+    if (g_root_island) {
+      g_root_island->prev = pi;
+    }
+    g_root_island = pi;
+    gpr_mu_unlock(&g_island_mu);
+  }
+}
+
+static void global_fork_pi_list_remove(polling_island *pi) {
+  if (grpc_fork_support_enabled()) {
+    gpr_mu_lock(&g_island_mu);
+    if (pi->prev) {
+      pi->prev->next = pi->next;
+    }
+    if (pi->next) {
+      pi->next->prev = pi->prev;
+    }
+    if (g_root_island == pi) {
+      g_root_island = g_root_island->next;
+    }
+    gpr_mu_unlock(&g_island_mu);
+  }
+}
+
+static void fd_register_postfork_handler(grpc_fd *fd,
+                                         fd_postfork_handler handler,
+                                         void *arg) {
+  fd->postfork_handler = handler;
+  fd->postfork_handler_arg = arg;
+}
+
+static void polling_island_reset(polling_island *pi) {
+  close(pi->epoll_fd);
+  pi->epoll_fd = epoll_create1(EPOLL_CLOEXEC);
+
+  struct epoll_event ev;
+  for (size_t i = 0; i < pi->fd_cnt; i++) {
+    ev.events = (uint32_t)(EPOLLIN | EPOLLOUT | EPOLLET);
+    ev.data.ptr = pi->fds[i];
+    GPR_ASSERT(epoll_ctl(pi->epoll_fd, EPOLL_CTL_ADD, pi->fds[i]->fd, &ev) ==
+               0);
+  }
+
+  grpc_error *error = GRPC_ERROR_NONE;
+  polling_island_add_wakeup_fd_locked(pi, &polling_island_wakeup_fd, &error);
+}
+
+static void fd_postfork(grpc_fd *fd) {
+  if (!fd->orphaned) {
+    if (fd->postfork_handler) {
+      fd->fd = fd->postfork_handler(fd, fd->postfork_handler_arg);
+    } else {
+      close(fd->fd);
```

Comment:
I am a bit unclear on this. Why isn't just doing a `close` enough ? - Why do we need to create a new socket and then `shutdown` ? 
Also, the `grpc_fd` will still be in the `polling_island` - is that the intention here ?